### PR TITLE
Cache results to not query on each call.

### PR DESCRIPTION
**Pull request by RomainLanz**

### CODE REVIEW

File: src/ExistingAttributesPersistence.php [hunk 1]
```diff
@@ -18,8 +18,11 @@ public static function bootExistingAttributesPersistence()
     {
         // Register once for all, the existing attributes of the table
         $dummy = new static;
-        static::$ExistingAttributesPersistence_tableColumns = \DB::connection()->getSchemaBuilder()->getColumnListing($dummy->getTable());
```
Comment (phurni) [inline src/ExistingAttributesPersistence.php:21]: The bootTRAITNAME() method should be called only once per model, so caching table columns here should not have an impact. Do you have some profling or log to confirm?

File: src/UserStamping.php [hunk 2]
```diff
@@ -2,11 +2,18 @@
 
 trait UserStamping {
 
+    /**
+     * Columns name.
+     *
+     * @var array
+     */
+    private $columns = [];
```
Comment (phurni) [inline src/UserStamping.php]: Could this (even private) attribute name clash with others, I think so. Please update it with the trait name as a prefix.

File: src/UserStamping.php [hunk 3]
```diff
@@ -65,11 +72,16 @@ public static function bootUserStamping()
...
 
     /**
-     * Add DB table columns meta info
+     * Gets columns name of model's table.
      *
+     * @return  array
      */
     protected function getUserStampingTableColumns() {
-        return \DB::connection()->getSchemaBuilder()->getColumnListing($this->getTable());
+        if ( ! $this->columns) {
```
Comment (phurni) [inline src/UserStamping.php]: Please update this line by removing the whitespaces in the condition, to be consistent with the rest of the code: `if (!$this->columns) {`

**Comment by RomainLanz** [comment]
Change done.

Concerning the `bootExistingAttributesPersistence()` method. As you can see bellow we win some queries.

Without condition:
capture d ecran 2015-02-13 a 17 00 46

With condition:
capture d ecran 2015-02-13 a 17 01 49

**Comment by phurni** [comment]
Let's merge this PR for UserStamping, please revert your changes for  `src/ExistingAttributesPersistence.php`, there are only whitespace changes which is a no-op.

Concerning the log, we should first investigate where the second query is triggered, maybe this is a dev environment only situation.

**Comment by RomainLanz** [comment]
All changes made in `src/ExistingAttributesPersistence.php` are reverted.

**Comment by RomainLanz** [comment]
What do you think of changing the `bootExistingAttributesPersistence()` method with something like this:

``` php
public static function bootExistingAttributesPersistence()
{
    // Listen to the 'saving' event but with a low priority so that any validation is performed before purging the attributes
    $name = get_called_class();
    static::getEventDispatcher()->listen("eloquent.saving: {$name}", function($model)
    {
        // Keep only the attributes that exists in the table
        $model->setRawAttributes(
            array_only(
                $model->getAttributes(),
                \DB::connection()->getSchemaBuilder()->getColumnListing($model->getTable())
            )
        );
    }, -100);
}
```

With that, the query is executed only when you save the model and not at every load.

Or maybe better to create another Trait named `StoreColumnsTrait` that store all columns in a static property. And all other Trait will depend on this. With that you have only one request per model for all your Trait.

``` php
<?php namespace Mortimer\Poignant;

trait ExistingAttributesPersistence {

    use StoreColumnsTrait;

    /**
     * Boot the trait for a model.
     *
     * @return void
     */
    public static function bootExistingAttributesPersistence()
    {
        // Listen to the 'saving' event but with a low priority so that any validation is performed before purging the attributes
        $name = get_called_class();
        static::getEventDispatcher()->listen("eloquent.saving: {$name}", function($model)
        {
            // Keep only the attributes that exists in the table
            $model->setRawAttributes(
                array_only(
                    $model->getAttributes(),
                    self::getDatabaseColumns()
                )
            );
        }, -100);
    }
}
```

``` php
<?php namespace Mortimer\Poignant;

trait StoreColumnsTrait {

    /**
     * All columns for the related model.
     *
     * @var array
     */
    private static $databaseColumns = [];

    /**
     * Gets columns name of the related model.
     * 
     * @return array
     */
    public static function getDatabaseColumns()
    {
        if (!self::$databaseColumns) {
            self::$databaseColumns = \DB::connection()->getSchemaBuilder()->getColumnListing((new static)->getTable())
        }

        return self::$databaseColumns;
    }

}
```

**Comment by phurni** [comment]
1. Yes, the `tablecolumns` should be memoized when they are used, not before. Your proposal doesn't memoize though. (Maybe in reference to 2.)
2. I had the same "vision" when developing these traits, I think (prove me wrong please) this is not possible because a trait can only be `use`'d once in a class. So multiple traits using another one (`StoreColumnsTrait`) in the same class will fail at compile time. This limitation certainly has to do with method precedence in the trait system.

**Comment by RomainLanz** [comment]
You are true. Maybe better to add theses functions on the Core of Laravel.

**Pull request merged**